substrate/frame/system/src/offchain.rs

```diff
@@ -457,38 +447,54 @@ pub trait SigningTypes: crate::Config {
 	type Signature: Clone + PartialEq + core::fmt::Debug + codec::Codec + scale_info::TypeInfo;
 }
 
-/// A definition of types required to submit transactions from within the runtime.
-pub trait SendTransactionTypes<LocalCall> {
-	/// The extrinsic type expected by the runtime.
-	type Extrinsic: ExtrinsicT<Call = Self::OverarchingCall> + codec::Encode;
+/// Common interface for the `CreateTransaction` trait family to unify the `Call` type.
+pub trait CreateTransactionBase<LocalCall> {
+	/// The extrinsic.
+	type Extrinsic: ExtrinsicLike + Encode;
+
 	/// The runtime's call type.
 	///
 	/// This has additional bound to be able to be created from pallet-local `Call` types.
-	type OverarchingCall: From<LocalCall> + codec::Encode;
+	type RuntimeCall: From<LocalCall> + Encode;
 }
 
-/// Create signed transaction.
-///
-/// This trait is meant to be implemented by the runtime and is responsible for constructing
-/// a payload to be signed and contained within the extrinsic.
-/// This will most likely include creation of `SignedExtra` (a set of `SignedExtensions`).
-/// Note that the result can be altered by inspecting the `Call` (for instance adjusting
-/// fees, or mortality depending on the `pallet` being called).
+/// Interface for creating a transaction.
+pub trait CreateTransaction<LocalCall>: CreateTransactionBase<LocalCall> {
+	/// The extension.
+	type Extension: TypeInfo;
+
+	/// Create a transaction using the call and the desired transaction extension.
+	fn create_transaction(
+		call: <Self as CreateTransactionBase<LocalCall>>::RuntimeCall,
+		extension: Self::Extension,
+	) -> Self::Extrinsic;
+}
+
+/// Interface for creating an old-school signed transaction.
 pub trait CreateSignedTransaction<LocalCall>:
-	SendTransactionTypes<LocalCall> + SigningTypes
+	CreateTransactionBase<LocalCall> + SigningTypes
 {
+	/// TODO: revisit this, currently here for metadata purposes.
+	type SignaturePayload: SignaturePayload;
+
 	/// Attempt to create signed extrinsic data that encodes call from given account.
 	///
 	/// Runtime implementation is free to construct the payload to sign and the signature
 	/// in any way it wants.
 	/// Returns `None` if signed extrinsic could not be created (either because signing failed
 	/// or because of any other runtime-specific reason).
-	fn create_transaction<C: AppCrypto<Self::Public, Self::Signature>>(
-		call: Self::OverarchingCall,
+	fn create_signed_transaction<C: AppCrypto<Self::Public, Self::Signature>>(
+		call: <Self as CreateTransactionBase<LocalCall>>::RuntimeCall,
 		public: Self::Public,
 		account: Self::AccountId,
 		nonce: Self::Nonce,
-	) -> Option<(Self::OverarchingCall, <Self::Extrinsic as ExtrinsicT>::SignaturePayload)>;
+	) -> Option<Self::Extrinsic>;
+}
+
+/// Interface for creating an inherent.
+pub trait CreateInherent<LocalCall>: CreateTransactionBase<LocalCall> {
```

Comment:
IMHO we should name it `CreateBareTransaction`.

I guess this will be removed when new extrinsics horizon is done.
But in the meantime it brings more confusion than help.

Currently unsigned extrinsics can be inherents or just unsigned transactions (like equivocations report).

Inherents are always created by `ProvideInherent` trait implementation. the call is marked as inherent with `ProvideInherent::is_inherent` implementation.
Those inherents are enforced to be first in the block before all other extrinsics (it is not the case for equivocations-style inherents)

But now this PR also calls all bare transactions inherent. So it is not consistent with the usage of `ProvideInherent::is_inherent` implementation.

Either we come up with a new name for provided inherents which must be first in the block etc... or we just call the transaction here "bare".

Same `UncheckedExtrinsic` has a function `is_inherent` which is in conflict with `is_inherent` from `ProvideInherent`.

I feel like sticking to `is_bare` is less confusing.

Reply:
In order to create a general transaction from a call and a transaction extension there is the trait `CreateTransaction`.

Here `CreateInherent` which IMHO should be named `CreateBare` is to create a bare extrinsic from a call only, without transaction extension.

Offchain worker are using this to generate extrinsics which later will be validate with `validate_unsigned`.

Also in a follow-up PR we want to introduce `pallet::authorize(|args| { checks })` as an attribute on top of calls, and the system transaction extension `AuthorizeCalls`, which will execute those checks and return a new origin `SystemOrigin::Authorized`.
So that people can migrate easily from `validate_unsigned` to this new logic.
I have a draft branch: gui-transaction-extension-stage-2

But this doesn't work for transaction with `TransactionSource` being local.

`TransactionExtension::validate` function doesn't give access to `TransactionSource`, contrary to `ValidateUnsigned::validate_unsigned` function.

IMO we might want to just put `TransactionSource` as an arg in `TransactionExtension::validate`.
Or we can try to get rid of the usage of `TransactionSource` but that means probably to make some stuff signed by block author in offchain worker. or maybe move transaction from local source to inherents (but this looks even harder).

Reply:
Yes we can open a follow-up issue and PR about transaction source in the transaction extension.

The main usecase I see is for code which you want don't want to pay, and you want to trust the offchain worker, and is compute-expensive. Like checking election solution for instance.

Anyway, refactoring without requiring transaction source could be doable. So a follow-up sounds best.